ci/ci/ci.py

```diff
@@ -121,13 +126,54 @@
 
     batch_client = request.app['batch_client']
     batches = batch_client.list_batches(
-        f'test=1 pr={pr_number}')
+        f'test=1 pr={pr.number}')
     batches = sorted([b async for b in batches], key=lambda b: b.id, reverse=True)
     page_context['history'] = [await b.status() for b in batches]
 
     return await render_template('ci', request, userdata, 'pr.html', page_context)
 
 
+async def retry_pr(wb, pr, request):
+    app = request.app
+    session = await aiohttp_session.get_session(request)
+
+    if pr.batch is None:
+        log.info('retry cannot be requested for PR #{pr.number} because it has no batch')
+        set_message(
+            session,
+            f'Retry cannot be requested for PR #{pr.number} because it has no batch.',
+            'error')
+        raise web.HTTPFound(
+            deploy_config.base_path('ci') +
+            f'/watched_branches/{wb.index}/pr/{pr.number}')
+    batch_id = pr.batch.id
+
+    dbpool = app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('INSERT INTO invalidated_batches (batch_id) VALUES (%s);', batch_id)
+
+    pr.batch = None
+    pr.build_state = None
```

Comment:
I think once you notify even these are unnecessary.

Reply:
Notify only sets `_batch_changed`. The first lines of `PR._update_batch` are:
```
    async def _update_batch(self, batch_client, dbpool):
        if self.build_state:
            assert self.batch
            return
```

Updating GitHub works differently. It always contacts GitHub for a state update.

Reply:
Ah, `_update_batch` is wrong now that the current batch can become not the current batch, which didn't happen before.  I think it is trying to hard to be clever, and should just unconditionally grab the latest batch for the PR from batch.